### ISSUE TITLE
Vault Generated Credentials 

### ISSUE THREAD

**Issue by bkendall** [hashicorp/consul-template#446]
I seem to be having trouble with getting generated (AWS) credentials into a template. I've been tweaking this part of the template quite a bit, but with no success. I'm wondering if I'm doing something silly or there's a known issue. The `vault read` command line works for this path, but it's not coming through in the template. Other vault values are working, but not these generated credentials. Any suggestions?

Template:
```
{{ with vault "aws/creds/read-tags" }}
AWS_ACCESS_KEY={{ .Data.access_key }}
AWS_SECRET_KEY={{ .Data.secret_key  }}
{{ end }}
```

Output:
```bash
AWS_ACCESS_KEY=<no value>
AWS_SECRET_KEY=<no value>
```

Versions:
```bash
» vault version
Vault v0.3.1
» consul-template --version
consul-template v0.11.0
» consul version
Consul v0.5.2
Consul Protocol: 2 (Understands back to: 1)
```

edit: vault CLI output formatted as json (keys redacted). things seem to be working...
```
» vault read -format=json aws/creds/read-tags
{
	"lease_id": "aws/creds/read-tags/dc1523f6-678c-8a55-01ac-995ebfdd2902",
	"lease_duration": 3600,
	"renewable": true,
	"data": {
		"access_key": "[access key]",
		"secret_key": "[secret key]"
	}
}
```

**Comment by rsandor** [comment]
:+1:

**Comment by sethvargo** [comment]
Hi @bkendall 

Thank you for opening an issue and I am sorry you are having issues with Vault + Consul Template. I just tried Consul Template v0.11.0 and it's working for me as intended:

Here is the script I am using to run Vault and mount the AWS creds and policy:

```bash
#!/usr/bin/env bash
set -e

# Get the parent directory of where this script is.
SOURCE="${BASH_SOURCE[0]}"
while [ -h "$SOURCE" ] ; do SOURCE="$(readlink "$SOURCE")"; done
DIR="$(cd -P "$(dirname "$SOURCE")" && pwd)"

# Change into that directory
pushd $DIR &>/dev/null

echo "==> Starting Vault..."
vault server -dev &>vault.log &
sleep 1
VAULT_PID=$!
export VAULT_ADDR=http://127.0.0.1:8200
echo "==> Vault is running ($VAULT_PID)"

# Stop vault when we stop the script
trap "kill $VAULT_PID" SIGINT SIGTERM EXIT

# AWS
echo "==> Setting up AWS backend..."
vault mount aws &>/dev/null
envchain aws /bin/bash <<EOH
  vault write aws/config/root \
    access_key="\$AWS_ACCESS_KEY_ID" \
    secret_key="\$AWS_SECRET_ACCESS_KEY" \
    region=us-east-1 &>/dev/null
EOH
vault write aws/config/lease \
  lease="1h" \
  lease_max="12h" &>/dev/null
echo "==> Creating AWS policy for developers..."
vault write aws/roles/developer \
  policy="@$PWD/policy.json" &>/dev/null

# All good!
echo "==> Ready!"

# Basically block so the server runs in the background
wait $VAULT_PID

echo "==> Cleaning up..."

popd
```

Here is the policy (although that really shouldn't matter):

```json
{
  "Version": "2012-10-17",
  "Statement": {
    "Action": "ec2:*",
    "Effect": "Allow",
    "Resource": "*",
    "Condition": {
      "Bool": { "aws:MultiFactorAuthPresent": true }
    }
  }
}
```

Here is the template:

```liquid
{{ with vault "aws/creds/developer" }}
AWS_ACCESS_KEY={{ .Data.access_key }}
AWS_SECRET_KEY={{ .Data.secret_key  }}
{{ end }}
```

Here is the command I ran:

```sh
VAULT_ADDR=http://127.0.0.1:8200 VAULT_TOKEN=e69b4b18-82a4-1eb2-7a77-ef154cb22010 consul-template -template in.ctmpl:out.txt
```

Could you please provide more information, including the debug logs?

**Comment by bkendall** [comment]
Alright, @sethvargo, took your script, modified a touch to work in my env, and got `<no value>` :sob: 

Here's some additional information:

First, environment: `os x 10.11.1`

Second, I made a [gist of all the logs](https://gist.github.com/bkendall/cf1c56e9a894a8b1c038), as to not completely pollute this discussion.

Interestingly, I found that we write a lease length to vault, but when template reads it, [it sees some issue with the lease length](https://gist.github.com/bkendall/cf1c56e9a894a8b1c038#file-consul-template-log-L66).

Let me know what you think or if you need more information from me.

**Comment by sethvargo** [comment]
@bkendall which change did you make to the script?

Thank you for getting me the logs, although they don't show anything useful sadly :frowning:.

**Comment by bkendall** [comment]
@sethvargo 

small changes to make it work locally (and be slightly more verbose while I was getting it running)

```bash
13c13
< vault server -dev &>vault.log &
---
> vault server -log-level=debug -dev &>vault.log &
25,30c25,28
< envchain aws /bin/bash <<EOH
<   vault write aws/config/root \
<     access_key="\$AWS_ACCESS_KEY_ID" \
<     secret_key="\$AWS_SECRET_ACCESS_KEY" \
<     region=us-east-1 &>/dev/null
< EOH
---
> vault write aws/config/root \
>   access_key=[my key] \
>   secret_key=[my key] \
>   region=us-west-2
33c31
<   lease_max="12h" &>/dev/null
---
>   lease_max="12h"
36c34
<   policy="@$PWD/policy.json" &>/dev/null
---
>   policy=@policy.json
```

is there any way I can get you more verbose logs?

**Comment by sethvargo** [comment]
@bkendall does the issue reproduce without editing the script at all? what does your policy look like?

**Comment by bkendall** [comment]
@sethvargo my policy looks like the following (much security):

```json
{
  "Version": "2012-10-17",
  "Statement": {
    "Action": "ec2:*",
    "Effect": "Allow",
    "Resource": "*"
  }
}
```

I had to remove `envchain`, since I don't have that, and I replaced the environment variables (and region) with ones that were applicable to me. Running `vault read` _does_ work for me, but the template still comes up empty... I'll see if I can get this tested in Ubuntu, just on the off chance that something in OS X is messing with me...

**Comment by bkendall** [comment]
FWIW: I have other `vault` values in my template that were working as well, but it was specifically the credentials that weren't coming through

**Comment by sethvargo** [comment]
@bkendall I'm using OS X locally... Can you try without once?

**Comment by bkendall** [comment]
@sethvargo same result, `<no value>`. [log file](https://gist.github.com/bkendall/6bf48ccd3b51897874fa) :/

**Comment by sethvargo** [comment]
I really don't have any ideas then...

**Comment by bkendall** [comment]
@sethvargo thanks for bouncing back and forth with me. I'm going to give it a try on a different box and see if it's just something in my setup, and I'll get back to you...

**Comment by bkendall** [comment]
@sethvargo it looks like it was fixed in `v0.11.1`. What's interesting is that there is no [lease length issue](https://gist.github.com/bkendall/cf1c56e9a894a8b1c038#file-consul-template-log-L66) like what existed before, so I imagine that this was somehow related to #443, #444.

Thanks for your help.
